libretro-common/net/net_http.c

```diff
@@ -662,167 +615,626 @@ void net_http_connection_free(struct http_connection_t *conn)
+         prev = current;
+      else
+      {
+         /* need to remove current */
+         if (prev)
+            prev->next = current->next;
+         else
+            conn_pool = current->next;
+      }
+      current = current->next;
+   }
+   prev->next = entry;
+   entry->next = NULL;
+}
+
+static struct conn_pool_entry *net_http_conn_pool_find(const char *domain, int port)
+{
+   struct conn_pool_entry *entry;
+
+#ifdef HAVE_THREADS
+   slock_lock(conn_pool_lock);
+#endif
+
+   net_http_conn_pool_remove_expired();
+
+   entry = conn_pool;
+   while (entry)
+   {
+      if (!entry->in_use && port == entry->port && string_is_equal(entry->domain, domain))
+      {
+         entry->in_use = true;
+         net_http_conn_pool_move_to_end(entry);
+#ifdef HAVE_THREADS
+         slock_unlock(conn_pool_lock);
+#endif
+         return entry;
+      }
+      entry = entry->next;
+   }
+#ifdef HAVE_THREADS
+   slock_unlock(conn_pool_lock);
+#endif
+   return NULL;
+}
+
+static struct conn_pool_entry *net_http_conn_pool_add(const char *domain, int port, int fd, bool ssl)
+{
+   struct conn_pool_entry *entry = (struct conn_pool_entry*)calloc(1, sizeof(*entry));
+   if (!entry)
       return NULL;
-#else
-   int fd                = -1;
-   if (!conn || (fd = net_http_new_socket(conn)) < 0)
+   entry->domain = strdup(domain);
+   entry->port = port;
+   entry->fd = fd;
+   entry->in_use = true;
+   entry->ssl = ssl;
+   entry->connected = false;
+#ifdef HAVE_THREADS
+   slock_lock(conn_pool_lock);
+#endif
+   net_http_conn_pool_move_to_end(entry);
+#ifdef HAVE_THREADS
+   slock_unlock(conn_pool_lock);
+#endif
+   return entry;
+}
+
+struct http_t *net_http_new(struct http_connection_t *conn)
+{
+   struct http_t *state;
+
+   if (!conn)
       return NULL;
+
+   state = (struct http_t*)calloc(1, sizeof(struct http_t));
+   if (!state)
+      return NULL;
+
+   state->ssl  = conn->ssl;
+   state->conn = NULL;
+
+   state->request.domain        = strdup(conn->domain);
+   state->request.path          = strdup(conn->path);
+   state->request.method        = strdup(conn->method);
+   state->request.contenttype   = conn->contenttype ? strdup(conn->contenttype) : NULL;
+   state->request.contentlength = conn->contentlength;
+   if (conn->postdata && conn->contentlength)
+   {
+      state->request.postdata   = malloc(conn->contentlength);
+      memcpy(state->request.postdata, conn->postdata, conn->contentlength);
+   }
+   state->request.useragent     = conn->useragent ? strdup(conn->useragent) : NULL;
+   state->request.headers       = conn->headers ? strdup(conn->headers) : NULL;
+   state->request.port          = conn->port;
+
+   state->response.status  = -1;
+   state->response.buflen  = 16 * 1024;
+   state->response.data    = (char*)malloc(state->response.buflen);
+   state->response.headers = string_list_new();
+   string_list_initialize(state->response.headers);
+
+   return state;
+}
+
+static void net_http_resolve(void *data)
```

Comment:
@warmenhoven  Any reason to use an void pointer as parameter in a static function?